compiler/rustc_errors/src/diagnostic_impls.rs

```diff
@@ -170,6 +171,36 @@ impl IntoDiagnosticArg for Level {
     }
 }
 
+#[derive(Clone)]
+pub struct DiagnosticSymbolList(Vec<Symbol>);
```

Comment:
nit: ideally this is generic over `T: Display`, and also, ideally there's an And version and an Or version

Reply:
honestly I _think_ the right path forward is to give the custom derive first-class support for vecs and special case them, because it can't use IntoDiagnosticArg anyway once we're using a real list formatter since there's no way to pass in the list formatter.

But this is still a positive step.

(I'm wondering what @davidtwco thinks)

Reply:
re: Generic over `T: Display` or `T: IntoDiagnosticArg`, yes that's better in the long run. However the `IntoDiagnosticArg` impl for `Symbol` Currently doesn't emit the `` ` ``delimiter that many diagnostic message expected. That would change a lot of code, so i'd prefer leave that to a follow work. Created #103422

Reply:
For accessing the list formatter i don't think it's hard, we can just store `tcx` inside the `DiagnosticSymbolList`, so it will have access to any necessary information.

Reply:
A proper list formatter would basically just need to live on the session/emitter alongside the fluent bundle stuff.